src/Package/Impl/Sql/Publish/SqlPublishOptionsDialogViewModel.cs

```diff
@@ -0,0 +1,306 @@
+            if (Settings.TargetType != PublishTargetType.Dacpac) {
+                CanPublish &= Targets?.Count > 0;
+            }
+        }
+
+        private void PopulateTargets() {
+            _coreShell.AssertIsOnMainThread();
+
+            switch (Settings.TargetType) {
+                case PublishTargetType.Dacpac:
+                    Targets = new List<string>();
+                    break;
+
+                case PublishTargetType.Project:
+                    PopulateProjectList();
+                    break;
+
+                case PublishTargetType.Database:
+                    PopulateConnectionsList();
+                    break;
+            }
+        }
+
+        private void PopulateConnectionsList() {
+            var connections = GetDatabaseConnections();
+            var index = -1;
+            _targetConnections = connections;
+            if (!string.IsNullOrEmpty(Settings.TargetDatabaseConnection)) {
+                var indices = _targetConnections.IndexWhere(c => c.ConnectionString.EqualsIgnoreCase(Settings.TargetDatabaseConnection));
+                index = indices.DefaultIfEmpty(-1).First();
+            }
+            Targets = _targetConnections.Select(c => c.Name).ToList();
+            SelectedTargetIndex = index >= 0 ? index : 0;
+            UpdateState();
+        }
+
+        private void PopulateProjectList() {
+            _targetProjects = GetDatabaseProjects();
+            if (_targetProjects.Count > 0) {
+                var index = -1;
+                if (!string.IsNullOrEmpty(Settings.TargetProject)) {
+                    var indices = _targetProjects.IndexWhere(x => x.EqualsIgnoreCase(Settings.TargetProject));
+                    index = indices.DefaultIfEmpty(-1).First();
+                }
+                Targets = _targetProjects;
+                SelectedTargetIndex = index >= 0 ? index : 0;
+            }
+        }
+
+        private IReadOnlyList<string> GetDatabaseProjects() {
+            var projectMap = new Dictionary<string, EnvDTE.Project>();
+            var solution = _pss.GetSolution();
+            var projects = new List<string>();
+            foreach (EnvDTE.Project project in solution.Projects) {
+                try {
+                    // Some projects throw 'not implemented'
+                    var projectFileName = project.FileName;
+                    if (!string.IsNullOrEmpty(projectFileName) && Path.GetExtension(projectFileName).EqualsIgnoreCase(".sqlproj")) {
+                        projects.Add(project.Name);
+                    }
+                } catch (NotImplementedException) { } catch (COMException) { }
+            }
+
+            if (projects.Count == 0) {
+                projects.Add(Resources.SqlPublishDialog_NoDatabaseProjects);
+            }
+            return projects;
+        }
+
+        private IReadOnlyList<DbConnectionData> GetDatabaseConnections() {
```

Comment:
Ah, ok

---
In reply to: [73198718](https://github.com/Microsoft/RTVS/pull/2135#discussion_r73198718) [](ancestors = 73198718)